Guide error handling (#504)
restatedev/documentation: docs/guides/error-handling.mdx

---
title: "Error Handling"
description: "Learn how to handle transient and terminal errors in your applications."
pagination_next: null
pagination_prev: null
---

import Admonition from "@theme/Admonition";
import {CodeWithTabs} from "../../src/components/code/code";

# Error Handling

> Restate handles retries for failed invocations. By default, Restate does infinite retries with an exponential backoff strategy.

Reviewer: ```suggestion
Restate handles retries for failed invocations. By default, Restate infinitely retries all errors with an exponential backoff strategy.
```

Seems important to underline "all errors" here, given that's how you start the next section

This guide helps you fine-tune the retry behavior for your use cases.

## Infrastructure errors (transient) vs. application errors (terminal)

In Restate, we distinguish between two types of errors: transient errors and terminal errors.

- **Transient errors** are temporary and can be retried. They are typically caused by infrastructure issues (network problems, service overload, API unavailability,...).
- **Terminal errors** are permanent and should not be retried. They are typically caused by application logic (invalid input, business rule violation, ...).

## Handling transient errors via retries

Restate assumes by default that all errors are transient errors and therefore retryable.
If you do not want an error to be retried, you need to specifically label it as a terminal error ([see below](#application-errors-terminal)).

Restate lets you configure the retry strategy at different levels: at the Restate-level (global) and at the run-block-level.

### At the Restate-Level (Global)

This defines the default retry policy that will be used for all invocations, unless overridden at the service-, or run-block-level.

> You can set the global retry policy in the [Restate Server configuration](/operate/configuration/server).

Reviewer: Generally, I'm not sure it's a good idea to tell people to set `max_` bounds on the retry policy in the invoker like this. Doing so kinda changes the semantics of the restate development model, essentially breaking the run to completion mentality that you can use when developing using restate.

I think it's fair if you want to describe it, but i think this deserves a warning section saying that putting a max bound effectively breaks the _run to completion_ premise of Restate.

More generally, I expect people changing the global retry policy only in terms of the exponential factor and interval, to tune it depending on how much the deployment infrastructure of services can handle.

I wouldn't expect people adding a `max_` bounds here in the global policy, but rather in the service level policy (which we don't have yet... Perhaps I can work on it today, let's see, as it completes the story here...)

Author: I will just remove it, I agree

max-attempts = "10" # max number of attempts before terminal error; default: unset (=infinite)
```

You can tune this policy to your needs. Note that all durations should follow the [humantime format](https://docs.rs/humantime/latest/humantime/fn.parse_duration.html).

> You can also use a **fixed-delay retry policy**:

Reviewer: Now that I look at this, I wonder if I should simplify this whole retry policy thing into just exponential retry policy anyway, removing the `type` thing....

restate-server
```

### At the Run-Block-Level

Handlers use run blocks to execute non-deterministic actions, often involving other systems and services (API call, DB write, ...).
These run blocks are especially prone to transient failures, and you might want to configure a specific retry policy for them.
Most Restate SDKs allow this:

<CodeWithTabs groupId={"sdk"}>
    ```ts !!tabs TypeScript
    CODE_LOAD::ts/src/guides/retries.ts?1
    ```

    ```python !!tabs Python
    CODE_LOAD::python/src/guides/retries.py?1
    ```

    ```java !!tabs Java
    CODE_LOAD::java/src/main/java/guides/RetryRunService.java?1
    ```

    ```kotlin !!tabs Kotlin
    CODE_LOAD::kotlin/src/main/kotlin/guides/RetryRunService.kt?1
    ```

    ```rust !!tabs Rust
    CODE_LOAD::rust/src/guides/retries.rs
    ```

</CodeWithTabs>

Note that these retries are coordinated and initiated by the Restate Server.
So the handler goes through the regular retry cycle of suspension and re-invocation.

If you set a maximum number of attempts, then the run block will fail with a TerminalException once the retries are exhausted.

<Admonition type="note" title="Other levels of retry policies">

> - Service-level retry policies are planned and will come soon.
> - Handler-level retry policies are not planned.

Reviewer: ```suggestion
Service-level retry policies are planned and will come soon.
```

    ```

    ```go !!tabs Go
    CODE_LOAD::go/guides/retries.go#catch
    ```

    ```rust !!tabs Rust
    CODE_LOAD::rust/src/guides/retries.rs#catch
    ```
</CodeWithTabs>

<Admonition type="note" title="Sagas with Restate">
    When you throw a terminal error, you might need to undo the actions you did earlier in your handler to make sure that your system remains in a consistent state.
    Have a look at our [sagas guide](/guides/sagas) to learn more.
</Admonition>

## Cancellations are Terminal Errors

You can cancel invocations via the [CLI](/operate/invocation#cancelling-invocations), UI and programmatically.
When you cancel an invocation, it throws a terminal error in the handler processing the invocation the next time it awaits a Promise or Future of a Restate Context action (e.g. run block, RPC, sleep,...; `CombineablePromise` in TypeScript, `Awaitable` in Java).
Unless caught, This terminal error will propagate up the call stack until it reaches the original caller.

Here again, the handler needs to have [compensation logic](/guides/sagas) in place to make sure the system remains in a consistent state, when you cancel an invocation.

> ## Timeouts to bound response times

Reviewer: I'm not sure this documentation bit is related to this guide TBH, it doesn't feel it belongs here, because it's an ops config parameter, and it has nothing to do with the development model. Perhaps it should rather go in the operate/registration page or similar.

Also the title itself is IMO misleading. This is not the response time, as you would observe it from a caller perspective. The rest of the document talks about transient and application error, and how one gets retried by restate and the other is not retried, thus observed by the caller. But these timeouts here aren't related to what the caller can observe, it's rather how the code internally will behave, that's why IMO doesn't belong here.

Author: what about calling it  `Timeouts between Restate and the service`?
The goal of a guide is providing an overcoupling overview on certain topics.

> Restate also has timeouts to bound the response time of a service invocation.

Reviewer: I don't think this is the correct description, this makes me think to something like an HTTP timeout that, as a caller, I can configure.

I think this should rather be described as "execution bounds" or something alike.

Author: what about "timeouts between Restate and the service"

There are two types of timeouts in Restate:

### Inactivity timeout

When the Restate Server does not receive a next journal entry from a running handler within the inactivity timeout, it will ask the handler to suspend.
This timer guards against stalled service/handler invocations. Once it expires, Restate triggers a graceful termination by asking the service invocation to suspend (which preserves intermediate progress).

> By default, the inactivity timeout is set to one minute.

Reviewer: I think the documentation should tell you why you want to increase it, and why you want to decrease it.

In the case of the inactivity timeout, I think it should be something like the following:

* Increase it when you can afford many idle HTTP requests, and you want to avoid too many replays, especially with invocations with long journals.
* Decrease it when you want to free up idle HTTP requests as much as possible, and you're ok with having too many replays, and your invocations generally progress fast and don't stay idle too long

### Abort timeout

This timer guards against stalled service/handler invocations that are supposed to terminate.
The abort timeout is started after the 'inactivity timeout' has expired and the service/handler invocation has been asked to gracefully terminate.
Once the timer expires, it will abort the service/handler invocation.

By default, the abort timeout is set to one minute.
This timer potentially interrupts user code.
If the user code needs longer to gracefully terminate, then this value needs to be set accordingly.

<Admonition type="note" title="Long-running run blocks">
    If you have long-running `ctx.run` blocks, you need to increase both timeouts to prevent the handler from terminating prematurely.
</Admonition>

> ### Configuring the timeouts

Reviewer: You can set those on a services basis too with `restate svc config edit <SVC_NAME>`

abort-timeout = "1m"
```

```shell
restate-server --config-file restate.toml
```

Both timeouts follow the [humantime](https://docs.rs/humantime/latest/humantime/) format.

Or set it [via environment variables](/operate/configuration/server#environment-variables), for example:
```shell
RESTATE_WORKER__INVOKER__INACTIVITY_TIMEOUT=5m \
  RESTATE_WORKER__INVOKER__ABORT_TIMEOUT=5m \
  restate-server
```

## Common patterns

These are some common patterns for handling errors in Restate:

### Sagas

Have a look at the [sagas guide](/guides/sagas) to learn how to revert your system back to a consistent state after a terminal error.
Keep track of compensating actions throughout your business logic and apply them in the catch block after a terminal error.

### Dead-letter queue
A [dead-letter queue (DLQ)](https://aws.amazon.com/what-is/dead-letter-queue/) is a queue where you can send messages that could not be processed due to errors.

You can implement this in Restate by wrapping your handler in a try-catch block. In the catch block you can forward the failed invocation to a DLQ Kafka topic or a catch-all handler which for example reports them or backs them up.

<details>

>     <summary>Catching failed invocations before handler execution starts</summary>
> 
>     Some errors might happen before the handler code gets invoked/starts running (e.g. service does not exist, request decoding errors in SDK HTTP server, ...).
>     Handle these as follows:

Reviewer: You can catch only request decoding errors.

```suggestion
    <summary>Catching request decoding errors</summary>

    By default Restate terminally fails all requests with a bad input payload with `400`. You can handle this case as follows:
```

    - In case the caller waited for the response of the failed call, the caller can handle the propagation to the DLQ.
    - If the caller did not wait for the response (one-way send), you would lose these messages.
        You could handle the decoding errors by letting the called handler take raw input and doing the decoding and validation itself.
        In this case, it would be included in the try-catch block which would do the dispatching:

        <CodeWithTabs groupId={"sdk"}>
            ```ts !!tabs TypeScript
            CODE_LOAD::ts/src/guides/retries.ts#raw
            ```

            ```python !!tabs Python
            CODE_LOAD::python/src/guides/retries.py#raw
            ```

            ```java !!tabs Java

            CODE_LOAD::java/src/main/java/guides/RetryRunService.java#raw
            ```

            ```kotlin !!tabs Kotlin
            CODE_LOAD::kotlin/src/main/kotlin/guides/RetryRunService.kt#raw
            ```

            ```go !!tabs Go
            CODE_LOAD::go/guides/retries.go#raw
            ```

            ```rust !!tabs Rust
            CODE_LOAD::rust/src/guides/retries.rs#raw
            ```
        </CodeWithTabs>

>         The other errors mainly occur due to misconfiguration of your setup (e.g. wrong service name, wrong handler name, forgot service registration...).

Reviewer: You cannot handle those atm
